pkg/controllermanager/controller/cloudprofile/cloudprofile_seed_control.go

```diff
@@ -0,0 +1,111 @@
+	"github.com/gardener/gardener/pkg/client/kubernetes"
+	"github.com/gardener/gardener/pkg/logger"
+	"github.com/gardener/gardener/pkg/utils/imagevector"
+	kutil "github.com/gardener/gardener/pkg/utils/kubernetes"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/client-go/tools/cache"
+)
+
+func (c *Controller) seedAdd(obj interface{}) {
+	key, err := cache.MetaNamespaceKeyFunc(obj)
+	if err != nil {
+		logger.Logger.Errorf("Couldn't get key for object %+v: %v", obj, err)
+		return
+	}
+	c.seedQueue.Add(key)
+}
+
+func (c *Controller) reconcileSeedKey(key string) error {
+	_, seedName, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+
+	seed, err := c.seedLister.Get(seedName)
+	if apierrors.IsNotFound(err) {
+		logger.Logger.Debugf("[CLOUDPROFILE SEED RECONCILE] %s - skipping because Seed has been deleted", key)
+		return nil
+	}
+	if err != nil {
+		logger.Logger.Infof("[CLOUDPROFILE SEED RECONCILE] %s - unable to retrieve object from store: %v", key, err)
+		return err
+	}
+
+	if err := c.seedControl.ReconcileSeed(seed, key); err != nil {
+		c.seedQueue.AddAfter(key, 15*time.Second)
+	}
+	return nil
+}
+
+// SeedControlInterface implements the control logic for reconciling Seed. It is implemented as an interface to allow
+// for extensions that provide different semantics. Currently, there is only one implementation.
+type SeedControlInterface interface {
+	ReconcileSeed(seed *gardenv1beta1.Seed, key string) error
+}
+
+// NewDefaultSeedControl returns a new instance of the default implementation ControlInterface that
+// implements the documented semantics for CloudProfiles.
+func NewDefaultSeedControl(k8sGardenClient kubernetes.Interface, k8sGardenInformers gardeninformers.Interface, secrets map[string]*corev1.Secret, imageVector imagevector.ImageVector, identity *gardenv1beta1.Gardener, backupBucketLister gardencorelisters.BackupBucketLister) SeedControlInterface {
+	return &defaultSeedControl{k8sGardenClient, k8sGardenInformers, secrets, imageVector, identity, backupBucketLister}
+}
+
+type defaultSeedControl struct {
+	k8sGardenClient    kubernetes.Interface
+	k8sGardenInformers gardeninformers.Interface
+	secrets            map[string]*corev1.Secret
+	imageVector        imagevector.ImageVector
+	identity           *gardenv1beta1.Gardener
+	backupBucketLister gardencorelisters.BackupBucketLister
+}
+
+func (c *defaultSeedControl) ReconcileSeed(obj *gardenv1beta1.Seed, key string) error {
+	_, err := cache.MetaNamespaceKeyFunc(obj)
+	if err != nil {
+		return err
+	}
+
+	var (
+		seedObj    = obj.DeepCopy()
+		seedLogger = logger.NewFieldLogger(logger.Logger, "cloudprofile seed", seedObj.Name)
+		ctx        = context.TODO()
+	)
+
+	cloudProfile := &gardenv1beta1.CloudProfile{}
+	if err := c.k8sGardenClient.Client().Get(ctx, kutil.Key(seedObj.Spec.Cloud.Profile), cloudProfile); err != nil {
```

Comment:
[Basic idea behind keeping `ProviderProfile.spec.backup.region` optional](https://github.com/gardener/gardener/blob/master/docs/proposals/02-backupinfra.md#workflow) was as follows. We want to  optimise the backup network latency when we have backup object store provider same as the cloud provider of seed.  In such scenario,  use common `ProviderProfile.spec.backupProfile` ( basically common credentials) but region from seed to locate bucket near to seed. This way  backupBucket has  soft dependency on seed.  I know we want to have seed independent of cloudProfile which will be the case still. So, considering this i can think of following option:

1. Make the `ProviderProfile.spec.backup.region` mandatory field. 
    - Losses/Scrap the network latency optimisation thought mentioned above.
    - Removes soft dependency on seed.
2. Extension of above options
    - To bring in latency optimisation, one has to create multiple `ProviderProfile` with different `spec.backup.region`. 
    ```YAML
     # ProviderProfile.yaml
    seedSelector:
        type: <same-as-spec.backup.provider>
        region: <same-as-spec.backup.provider>
    ```
3. In case of `ProviderProfile.spec.backup.region` is optional, then should have `ProviderProfile.Spec.seedSelector` as (mandatory via validation check): 
    ```YAML
    # ProviderProfile.yaml 
    seedSelector:
        type: <same-as-spec.backup.provider>
    ```
    - And region will be taken from annotation, `seed.core.gardener.cloud/region`.
    - And make shoot scheduler intelligent, to opt out seed selected under ProviderProfile from Shoot without above annotation.
    - *At first look, this might look hacky but i think its fine work around to consider while taking decision*. 
4. **My Recommendation** How about introducing/keeping `spec.Cloud` section in seed.
    ```YAML
    #seed yaml
    spec:
      provider: #This is only metadata about the cloud provider of the seed and has no relation with core/ProviderProfile
        type: <some-provider-type> #{aws,azure, etc}
        region: <some-region> 
    ```
    - No change in current implementation. Seed spec will have this field mandatory.
    - This makes the "region" handling code in above option cleaner.
    - If i'm not wrong, the blocking cloud provider  *metaservices* via network policies also requires this knowledge. 
 
cc @amshuman-kr , @vlerenc

Reply:
I like 4. Topology information is required. Making it explicit makes sense.

Reply:
The problem is, and that’s why we we want to change the reference from seed->profile now with the new API, that a seed can support multiple profiles. Having a 1:1 mapping doesn’t make much sense. Generally, the profiles were only meant for shoots, but now they contain mixed information for both seed and shoot.

Would it make sense to move `backup` configuration from profile spec to seed spec? 

```yaml
kind: Seed
…
spec:
  …
  backup:
    provider:
      type: <some-name> # may be different then `.spec.provider.type`
    # region: <some-region> # optional, if not given then `.spec.provider.region` will be used
      secretRef:
        name: backup-secret
        namespace: garden
```

You can still optimize backup network latency:
* If no region is given then the same as `provider.region` is used
* If region is given then simply use this (because that’s what the user/administrator wants/configures)

I think what is described [here](https://github.com/gardener/gardener/blob/master/docs/proposals/02-backupinfra.md#workflow) can still be kept.
The only price is that the backup configuration has to be configured per seed, but IMHO it makes anyways more sense having it there.

WDYT?

Reply:
,I'm fine with it. keeping it on seed. As per @vlerenc's suggestion  Initially plan was to have it unique across cloudProfile, so we moved it there.

 But now thinking over your proposal, doesn't makes sense to keep region optional.  Cause in that case, the point
> If no region is given then the same as `provider.region` is used
 doesn’t make sense. There isn't second source for region.
And said that now `seed.spec.backup` will be constant then why not simply create 
core.backupBucket resource in garden cluster as a part of landscape setup and have reference of it in seed. i.e.
`seed.spec.backup: backupBucketName`.

Shoot controller will pick the backupBucket from seed while creating backupEntry for shoot.

Reply:
Also, i'll introduce the "shoot.spec.takeBackup bool" field in shoot spec to have no backup case.

Reply:
> doesn’t make sense.

Why not? If I use another provider for backup different from seed provider then I might need the region.

> as a part of landscape setup

Please not, that makes it more difficult for everybody installing Gardener properly.

> Also, i'll introduce the "shoot.spec.takeBackup bool" immutable field in shoot spec to have no backup case.

Why in the shoot spec? A user should not interfer with the backups, that's why we deprecated `.spec.backup` in the current shoot spec.  Let's not add it?

Reply:
Since I was mentioned, the primary goal was to have fewer buckets, but the secondary goal was to prepare for control plane migration and not having to move around backup data or rewire a lost/migrated control plane to a new bucket. The idea was to at least move a shoot cluster control plane freely within a region as the blob stores also store the data per region. Whether or not that information (the regional bucket) comes from a separate resource or is explicitly associated with a seed is your call. In either case, a future control plane migration controller must be certain that the restore controller (backup & restore sidecar today) will find the backup data. That can be by contract or, more explicitly, by referencing the backup bucket per seed. In the latter case, we need some logic to tie a seed to a backup bucket, but which seed is eligible is then only a matter of simple lookup of that information (shoot X was on seed A and seeds B and C share the same backup bucket as A, so B and C are all eligible to host cluster X in the future). That way, cross-infrastructure or cross-region or any other form of partitioning can be implemented at the point in time when the seeds are created, also manually. However, we prefer means to let Gardener take care of this association, when it `shoot`s the seeds, because we also plan eventually to make Gardener smart and add or remove seeds as needed, i.e. the set of seeds won't stay fixed.

Not sure whether that was helpful, because I have not followed the entire discussion @swapnilgm @rfranzke.

Reply:
OK, thanks, I think this problem ("move shoots from one seed to another without 'losing' the backup configuration) is covered by the general approach, right @swapnilgm? I don't think it's bound to how the backup is configured in the garden cluster.

The required configuration etcd-backup-restore is injected via the control plane webhook, and if I understand correctly this webhook gets its information solely out of the `BackupBucket/BackupEntry` extension resource.

Still, I guess I don't have the full picture. @swapnilgm, can you please explain whether control plane migration to another seed would still work, and if so, how it would look like?

Reply:
> Why not? If I use another provider for backup different from seed provider then I might need the region. 

 Of-course for cross provider does need to mention region. But, how do you decide its cross provider or same provider? -> As perGEP-4, we are going completely cloud agnostic for seed. We no more know the provider and region of the seed. So,that's why i mentioned having  `region` field optional on "seed.spec.backup" doesn't make sense.

> Please not, that makes it more difficult for everybody installing Gardener properly.

I don't think it will be difficult this way. But still as you wish. I'll go with your approach. 

> Why in the shoot spec? A user should not interfer with the backups, that's why we deprecated .spec.backup in the current shoot spec. Let's not add it?

i'm fine with not having it. We again thought of introducing it because, some how reconciler should know whether to configure backup for etcd or not. Let's say test-clsuter shoot don't need to have backup. In that case, the shoot creater has to assign it to `seed`(considering we are going ahead with backup config on seed) without backup. Which is again kind of indirect control to shoot creator. So, thought of making it direct so that shoot creator doesn't have to know whether any seed is configured with backup or not. AS i said i'm fine with such indirection. Let's not add it.

> ,Not sure whether that was helpful, because I have not followed the entire discussion 

Thats was indeed helpful to have more clarity of thoughts. Thanks

> I think this problem ("move shoots from one seed to another without 'losing' the backup configuration) is covered by the general approach, right  
Yes. control plane migration is taken in consideration with high priority while designing this, Once `core.backupEntry` is created which will have `core.backupBucket` associated with it. And it won't be changed even if shoot control plane is migrated to another one. So, there we won't be requiring any backup migration at all. This will also not affected even if `extension.backupBucket` extension is rescheduled on another seed in case of seed unavailability. (context: https://github.com/gardener/gardener/pull/1128#discussion_r313281792). This reminds me i have to add couple of more checks there. 

I'll update the PR with seed having backup configuration.

Reply:
> As perGEP-4, we are going completely cloud agnostic for seed. We no more know the provider and region of the seed. 

I don't understand, of course we know. We have to know. See https://github.com/gardener/gardener/pull/1284/files#diff-bf2774d9954baab517306db45a5b80bbR241-R243.

> I don't think it will be difficult this way. But still as you wish. I'll go with your approach.

Thanks. There is already a lot that needs to be done after Gardener is installed (creating `CloudProfile`s, `Seed`s, etc.), so if we can avoid another such step let's better do it.

> i'm fine with not having it. We again thought of introducing it because, some how reconciler should know whether to configure backup for etcd or not. Let's say test-clsuter shoot don't need to have backup. In that case, the shoot creater has to assign it to seed(considering we are going ahead with backup config on seed) without backup. Which is again kind of indirect control to shoot creator. So, thought of making it direct so that shoot creator doesn't have to know whether any seed is configured with backup or not. AS i said i'm fine with such indirection. Let's not add it.

Can we omit this use-case for now and focus on implementing shared-bucket approach? Let's think about this later (whether we want to support it (because today we don't), and if yes, how). OK?

> Yes. control plane migration is taken in consideration with high priority while designing this

Great, thank you for clarifying! :)

Reply:
> I don't understand, of course we know. We have to know. See https://github.com/gardener/gardener/pull/1284/files#diff-bf2774d9954baab517306db45a5b80bbR241-R243.

Ohh sorry. Thank you for pointing that out. My bad. All this discussion, i as missing this part that seed still has provider section. Probably because of the statement "According to this proposal their will not be a single profile for a seed" in first comment, i got confused little. It all makes sense now. Sorry again.

Reply:
np :)